scripts/reliability_report.py

```diff
@@ -0,0 +1,267 @@
+
+"""
+
+import argparse
+import asyncio
+import datetime
+import json
+import logging
+import os
+import time
+from typing import cast
+from collections import OrderedDict, defaultdict
+from datetime import datetime
+
+import jinja2
+import toml
+from google.cloud.bigtable.data import (
+    BigtableDataClientAsync,
+    ReadRowsQuery,
+    RowRange,
+)
+from google.cloud.bigtable.data.row_filters import FamilyNameRegexFilter
+
+RELIABILITY_FAMILY = "reliability"
+
+"""
+Reliability data is stored in bigtable under the `reliability_id` key.
+The `reliability_id` is an opaque string that can be passed from an outside source,
+but by default it's a UUID4() string, so we can make some presumptions.
+For now, we can presume that the string will follow most ASCII formats, so all characters
+will be between `/` and `{`. Added bonus is that bigtable will allow us to do "partial"
+lookups, so we can fetch all by looking at the range.
+"""
+
+
+class BigtableScanner:
+    """Bigtable access class"""
+
+    def __init__(self, log: logging.Logger, settings: argparse.Namespace):
+        try:
+            self.bigtable = BigtableDataClientAsync(
+                project=settings.bigtable["project"]
+            )
+            self.log = log
+            self.settings = settings
+        except Exception as e:
+            log.error(e)
+            raise
+
+    async def collect(self):
+        """Collect the data elements. (NOTE: This does a table scan. )"""
+        table = self.bigtable.get_table(
+            self.settings.bigtable.get("instance"),
+            self.settings.bigtable.get("table"),
+        )
+
+        # The key is a semi-opaque string value that may be closely related to a
+        # UUID4. This was chosen because time based keys can cause issues with
+        # bigtable.
+        # We can potentially break this into sub queries, but for now, treat
+        # it as a partial string that encompasses all values.
+        rrange = RowRange(start_key="/", end_key="{")
+        rfilter = FamilyNameRegexFilter(self.settings.track_family)
+        rrquery = ReadRowsQuery(row_ranges=rrange, row_filter=rfilter)
+
+        raw_rows = await table.read_rows(query=rrquery)
+        result = {}
+        # Meta information about each record
+        max_time = mean_time = 0
+        min_time = 1000
+        start_time = time.time()
+        milestone_log = defaultdict(lambda: dict(count=0, total_time=0, average_time=0))
+
+        # Meta information about all records
+        success_count = fail_count = row_count = expired_count = 0
+
+        # Iterate over the rows and collect interesting bits of info
+        for row_count, row in enumerate(raw_rows, start=1):
+            milestones = OrderedDict()
+            successful = expired = False
+            total_time = 0
+            for cell in row.cells:
+                milestone = bytearray(cell.qualifier).decode()
+                timestamp = int.from_bytes(cell.value) * 0.000000001
```

Comment:
Also I'm confused by the timestamp value: we're storing our `ms_since_epoch` in the cell value,  -- wouldn't it be `* 1000` for comparison against python's `time.time`? 

Since we're using the bigtable recognized/standard 64 bit be type, optionally you could use `int(cell)` which they have for the conversion from bytes.

```suggestion
                timestamp = int.from_bytes(cell.value) * 0.000000001
```

Reply:
This is pulling the time value stored in bigtable, which is a millisecond value stored as a u64 int vale.  We need to convert the integer to an appropriately scaled float to represent the milliseconds as the remainder. Currently, the UTC is expressed as a 10 digit number (e.g. `1741896409`). Multiplying this by the 10 digit `0.000000001` offsets the decimal point such that the remainder of the int64 value can provide the millisecond value compatible with python. Multiplying the u64 value by 1000 would only increase the integer base and would not resolve the remainder.

I'm not sure what the suggested change entails, since it appears that it is the same code. Is there some white-space difference that is not apparent?

Reply:
oops I meant .001 (not `*1000`) as python's `time.time` is seconds (w/ more precision in the fractional part) - this is now resolved